### PR TITLE
move to the new error response

### PR DESCRIPTION
**Pull request by chipaca**
daemon, client, docs/rest.md, snapd integration tests: move to the new error response

This implements the error response we agreed in Gramado, both for the sync error result and the async failure output, with human-readable message and optional machine-readable kind and value fields.

### CODE REVIEW

**Comment by mvo5** [comment]
LGTM

File: client/client.go [hunk 1]
```diff
@@ -121,13 +121,13 @@ type response struct {
 }
 
 // errorResult is the real value of response.Result when an error occurs.
-// Note that only the 'Str' field is unmarshaled from JSON representation.
+// Note that only the 'message' field is unmarshaled from JSON representation.
 type errorResult struct {
-	Str string `json:"str"`
+	Message string `json:"message"`
```
Comment (pedronis) [inline client/client.go]: @chipaca any reasons not to already introduce kind and value here?

File: daemon/response.go [hunk 2]
```diff
@@ -150,7 +151,7 @@ func ErrorResponse(status int) ErrorResponseFunc {
 		Status: status,
 	}
 
-	return r.SetError
+	return r.setError
```
Comment (pedronis) [inline daemon/response.go]: didn't come from this branch but I find it strange that ErrorResponse follow the *Response naming pattern but doesn't actually produce a response

File: client/client.go (hunk 1 above)
Comment (chipaca) [inline client/client.go]: None beyond that we aren't using it for anything yet in the client itself. When the client supports installing things, including the license agreement thing, it'll have to grow the right bits.

**Comment by pedronis** [comment]
@chipaca looks good to me, have a couple questions above

File: daemon/response.go (hunk 2 above)
Comment (chipaca) [inline daemon/response.go]: Ah. It used to be a `Response`. Can you suggest a different name? Before, `NotFound` was a `Response` (an `ErrorResponseFunc` was a `Response`), and `NotFound(err, "blah")` was a `Response` also; now only the latter is true.

Comment (pedronis) [inline daemon/response.go]: ErrorResponder ?  ErrorResponseMaker? it seems it's actually never used out response.go itself so not sure it needs a super cute name

Comment (pedronis) [inline daemon/response.go]: @chipaca or also MakeErrorResponder ?

Comment (chipaca) [inline daemon/response.go]: how about this

**Comment by pedronis** [comment]
@chipaca last name changes looks reasonable, I have tweaked a bit your description at the top if you don't mind.

+1 from me

**Pull request merged**